File: Python/ByRound/1497/1497_E2_Square_Free_Division_hard_version.py
# Problem: CF 1497 E2 - Square-Free Division (hard version)
# https://codeforces.com/contest/1497/problem/E2

"""
Algorithm: Dynamic Programming with Sliding Window

This solution addresses the problem of dividing an array into the minimum number of continuous segments such that no two numbers in the same segment have a product that is a perfect square. We are allowed to change at most k elements in the array to achieve this.

Key ideas:
1. For each number, reduce it to its square-free form (remove all square factors).
2. Use a sliding window technique to determine the minimum number of segments needed for a fixed number of changes.
3. Apply dynamic programming to find the optimal number of segments considering up to k changes.

Time Complexity: O(n * k * log(max(a_i)) + n * k^2)
Space Complexity: O(n * k)

The algorithm works by:
- Precomputing the square-free representation of numbers
- Using sliding window to find optimal segment boundaries for varying numbers of changes
- Using DP to combine results optimally

"""

MAXN = 10**7 + 5
INF = 10**18

# Precompute the square-free representation of each number up to MAXN
to = [i for i in range(MAXN)]
for i in range(2, 10**4):
    for j in range(i * i, MAXN, i * i):
        while to[j] % (i * i) == 0:
            to[j] //= i * i

_t = int(input())


def read():
    return [int(x) for x in input().split()]


s = [0] * MAXN
for __ in range(_t):
    n, k = read()
    a = [0] + read()
    # Transform array elements to their square-free forms
    a = [to[a[i]] for i in range(0, n + 1)]

    # f[i][ct]: minimum starting index for a segment ending at i with ct changes allowed
    f = [[0] * (k + 1) for _ in range(n + 2)]
    for ct in range(k + 1):
        j, now = 1, 0
        for i in range(1, n + 1):
            s[a[i]] += 1
            if s[a[i]] >= 2:
                now += 1
                if now > ct:
                    # Shrink window until we're within allowed changes
                    while now > ct:
                        now -= s[a[j]] >= 2
                        s[a[j]] -= 1
                        j += 1
            f[i][ct] = j
        for x in a:
            s[x] = 0

    # dp[i][j]: minimum segments for first i elements with j changes used
    dp = [[INF] * (k + 1) for _ in range(n + 2)]
    for j in range(0, k + 1):
        dp[0][j] = 0
    for i in range(1, n + 1):
        for j in range(k + 1):
            for p in range(j + 1):
                # Try using p changes for previous segment
                dp[i][j] = min(dp[i][j], dp[f[i][p] - 1][j - p] + 1)

    print(min(dp[n][0:]))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1497/1497_E2_Square_Free_Division_hard_version.cpp
// Problem: CF 1497 E2 - Square-Free Division (hard version)
// https://codeforces.com/contest/1497/problem/E2

/*
 * Problem: E2. Square-Free Division (hard version)
 *
 * Algorithm/Techniques:
 * - Preprocessing: Reduce each number to its square-free form by removing all
 * square factors.
 * - Sliding window: For each number of operations k, find the minimum number of
 * segments required.
 * - Dynamic Programming: Use DP to compute minimum segments with up to k
 * changes.
 *
 * Time Complexity: O(n * k * sqrt(max(a[i])) + n * k^2)
 * Space Complexity: O(n * k + max(a[i]))
 */

#include <bits/stdc++.h>
using namespace std;

const int N = 2e5 + 10, K = 25, S = 1e7 + 10;
const int inf = 0x3f3f3f3f;
int T, n, k;
int ans;
int a[N];
int num[S];
int f[N][K], g[N][K];

inline int read() {
  int x = 0, f = 1;
  char ch = getchar();
  while (ch < '0' || ch > '9')
    ch == '-' ? f = 0 : 0, ch = getchar();
  while (ch >= '0' && ch <= '9')
    x = (x << 3) + (x << 1) + (ch ^ 48), ch = getchar();
  return f ? x : -x;
}

inline int min(int a, int b) { return a < b ? a : b; }

int main() {
  T = read();
  while (T--) {
    n = read(), k = read(), ans = inf;
    for (int i = 1; i <= n; ++i) {
      a[i] = read();
      // Reduce each number to its square-free form
      for (int j = 2; j * j <= a[i]; ++j)
        while (a[i] % (j * j) == 0)
          a[i] /= j * j;
    }

    // Compute for all possible numbers of operations (0 to k)
    for (int i = 0; i <= k; ++i) {
      for (int j = 1; j <= n; ++j)
        num[a[j]] = 0;
      // Sliding window: for each window with at most i duplicates, find the
      // leftmost valid start
      for (int l = 1, r = 1, t = 0; r <= n; ++r) {
        ++num[a[r]], t += (num[a[r]] > 1);
        while (t > i && l < r)
          t -= (num[a[l]] > 1), --num[a[l]], ++l;
        g[r][i] = l;
      }
    }

    // Dynamic Programming: f[i][j] represents the minimum segments needed for
    // first i elements with j operations
    for (int i = 1; i <= n; ++i)
      for (int j = 0; j <= k; ++j) {
        f[i][j] = inf;
        for (int x = 0; x <= j; ++x)
          f[i][j] = min(f[i][j], f[g[i][x] - 1][j - x] + 1);
      }

    // Find minimum answer among all possible operations from 0 to k
    for (int i = 0; i <= k; ++i)
      ans = min(ans, f[n][i]);

    printf("%d\n", ans);
  }
  return 0;
}

// https://github.com/VaHiX/CodeForces/